kimchi/src/circuits/polynomials/foreign_mul/gate.rs

```diff
@@ -0,0 +1,528 @@
+//! Foreign field multiplication gate
+
+use ark_ff::{FftField, Zero};
+use ark_poly::{univariate::DensePolynomial as DP, Evaluations as EV, Radix2EvaluationDomain as D};
+use array_init::array_init;
+use rand::{prelude::StdRng, SeedableRng};
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+
+use crate::{
+    alphas::Alphas,
+    circuits::{
+        argument::{Argument, ArgumentType},
+        constraints::ConstraintSystem,
+        domains::EvaluationDomains,
+        expr::{self, Column, E},
+        gate::{CircuitGate, GateType},
+        polynomial::COLUMNS,
+        wires::GateWires,
+    },
+    proof::ProofEvaluations,
+};
+
+use super::{ForeignMul0, ForeignMul1, ForeignMul2};
+
+impl<F: FftField> CircuitGate<F> {
+    /// Create foreign multiplication gate
+    pub fn create_foreign_mul(wires: &[GateWires; 8]) -> Vec<Self> {
```

Comment:
two things:

* take the current `row: usize` as argument instead of wires, and return `(usize, Vec<Self>)` where `usize` is the next `row` that the caller can use to continue building their circuit (see poseidon)